proposals/interceptors-issues-2024-01.md

```diff
@@ -0,0 +1,147 @@
+# Open issues for interceptors
+
+## File path portability
+
+One of our design goals with interceptors was to have a particular form of "portability": essentially, we should be able to run the source generators for a project, commit the results to disk, then remove the source generators from the project and be able to build in a variety of environments. Note that this is different from *determinism*--we don't mind if the resulting assembly differs in things like layout, etc. in each build environment, but want it to be the case that if an interceptor declaration functions in one environment, then it also functions in a different environment, assuming all the sources and dependencies are fully available in each environment.
+
+However, we currently require absolute paths for `InterceptsLocationAttribute`, with pathmap substitution performed if present. This doesn't work across local builds and CI because only CI will actually pass `/pathmap` during the build. The local build prefers to preserve local paths (e.g. in the PDB) so that the local versions of the sources can be discovered easily during debugging. The local and CI environments will have different answers for what they expect the paths to be, and the InterceptsLocation will have errors in one of the two environments.
+
+Having looked at several strategies to address here, including adjusting when the build system passes `/pathmap` to the compiler, it seems the most straightforward way is to permit relative paths. This is a strategy which is already proven out by `#line` directives.
+
+When a relative path appears in an `[InterceptsLocation]`, we will resolve it relative to the path of the containing file.
+
+```cs
+// /project/src/file1.cs
+class Original
+{
+    static void Interceptable() { }
+
+    void M()
+    {
+        Interceptable();
+    }
+}
+
+// /project/obj/generated/file2.g.cs
+class Interceptors
+{
+    [InterceptsLocation("../../src/file1.cs", ...)] // resolves to "/project/src/file1.cs"
+    public static void Interceptor() { }
+}
+```
+
+This means that generator authors will be best advised to produce relative paths using e.g. `Path.GetRelativePath("/project/src/file1.cs", "/project/obj/generated/file2.g.cs")`.
+
+Clearly this means there is a need for a generated file to know what path it "would be written out" to, while the text is being generated, before calling `AddSource` with a `hintName`. We may want to add a new public API like the following for this:
+
+```diff
+ namespace Microsoft.CodeAnalysis;
+
+ public readonly struct SourceProductionContext
+ {
+     public void AddSource(string hintName, string source);
++    public string GetFilePath(string hintName);
```

Comment:
Think we need to add a bit more about what this does. In my mind this returns the path the file would be written out to _if_ `$(EmitCompilerGeneratedFiles)` is `true`. Also I think we need to firmly state that `SyntaxTree` produced from generated files will effectively, or actually, have this set as the path.